[WebDriver][BiDi] Implement core script.evaluate execution, validation, and serialization

gitstart opened the pull request:
#### ce6c9a6fceafc8c5678966583f6162bd2c175ced
<pre>

This patch introduces the first milestone of the WebDriver BiDi script.evaluate implementation. It
includes parameter validation, target resolution, JavaScript execution, exception handling, and
remote value serialization according to the BiDi specification.

</pre>

## Summary

- Evaluates JavaScript using `globalThis.eval()` in both browsing-context and realm targets, with
proper sandbox validation
- Validates all incoming parameters (expression, target, resultOwnership, serializationOptions)
- Reports exceptions through BiDi ExceptionDetails, with stack parsing for both JSC and V8 formats
- Implements W3C BiDi remote value serialization algorithm for primitives, objects, arrays, Map/Set, Date, RegExp, special numeric values, cyclic structures, and depth limiting
- Produces deterministic key ordering for stable serialization output

## Technical Approach

- Envelope pattern (`{ __ok, remote }` / `{ __ok, e }`) for unified success/error handling between
JavaScript and C++
- Reusable exception helpers (`parseStackTrace()`, `buildExceptionDetailsFromExceptionValue()`,
`createExceptionDetails()`) designed for future commands
- Temporary pre-dispatch validation for `resultOwnership` enum parameter until protocol generator supports optional enum validation

## Key Files Touched

### UIProcess

**BidiScriptAgent.cpp/h**

- Implements the command handler, result processing, and serialization helpers
- Key functions: `evaluate()`, `finishEvaluateBidiScriptResult()`, `serializeAsRemoteValue()`,
`deserializeRemoteValue()`, `parseStackTrace()`, `buildExceptionDetailsFromExceptionValue()`,
`createExceptionDetails()`

**WebDriverBidiProcessor.cpp**

- Routes incoming messages to BidiScriptAgent and performs pre-dispatch validation
- Implements strict check for `resultOwnership` parameter and error code mapping

**WebAutomationSession.cpp/h**

- IPC bridge with async callback registry
- Key function: `evaluateBidiScript()`

### WebProcess

**WebAutomationSessionProxy.cpp/h**

- Receives the IPC request and executes the evaluation within the correct JavaScript realm
- Key function: `evaluateBidiScript()` - bridges C++ to JavaScript execution context

**WebAutomationSessionProxy.js**

- Performs the actual JavaScript execution and W3C BiDi remote value serialization
- Key functions: `evaluateBidiScript()`, `_executeBidiScript()`, `serializeBidiRemoteValue()`
- Implements cyclic-reference detection and depth control

### IPC

[**[WebAutomationSessionProxy.messages.in](http://webautomationsessionproxy.messages.in/)**](http://webautomationsessionproxy.messages.in/)

- Adds the new `EvaluateBidiScript` IPC message connecting UIProcess and WebProcess

### Protocol

**protocol/BidiScript.json**

- Updates RemoteValue definitions (bigint, boolean, number, string, undefined, null)
- Corrects the `includeShadowTree` field in SerializationOptions

## Tests

All tests in the following WPT files pass (100%):

- `primitive_values.py` (13/13) - All primitive types
- `strict_mode.py` (1/1) - Strict mode evaluation
- `target.py` (1/1) - Target resolution
- `invalid.py` (61/61) - Complete parameter validation coverage

**Overall: 119/295 script.evaluate tests passing (40.3%)**

Bug URL: [`https://bugs.webkit.org/show_bug.cgi?id=301159`](https://bugs.webkit.org/show_bug.cgi?id=301159)

No new tests (OOPS!).

* Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
(WebKit::convertJSONToRemoteValue):
(WebKit::BidiScriptAgent::evaluate):
(WebKit::BidiScriptAgent::finishEvaluateBidiScriptResult):
(WebKit::BidiScriptAgent::serializeAsRemoteValue):
(WebKit::BidiScriptAgent::RealmRegistryStub::realmIdForContext const):
(WebKit::BidiScriptAgent::RealmRegistryStub::contextForRealmId const):
* Source/WebKit/UIProcess/Automation/BidiScriptAgent.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::evaluateBidiScript):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::processBidiMessage):
(WebKit::toBidiErrorCode):
* Source/WebKit/UIProcess/Automation/protocol/BidiScript.json:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::evaluateBidiScript):
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.h:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.js:
(let.AutomationSessionProxy.prototype.evaluateBidiScript):
(let.AutomationSessionProxy.prototype._executeBidiScript):
(let.AutomationSessionProxy):
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.messages.in:
* WebDriverTests/TestExpectations.json:
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2586d6ce3079b2f83732319a49000c4153c45d14

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133029 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5530 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44127 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140565 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85062 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6044 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5393 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101731 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69051 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135975 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4260 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119204 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82530 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4145 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1695 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83800 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113215 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143218 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5199 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37902 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110110 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5281 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4457 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110292 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3995 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115465 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/58787 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5253 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33815 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5094 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68705 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5343 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5211 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

juliopiubello committed:
Description:

 - BiDi script.evaluate pipeline
      - Added UI→WP bridge to evaluate raw expressions with await/timeout and return an envelope. Files:
          - Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:2162
          - Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:489
          - Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.messages.in:26
          - Source/WebKit/UIProcess/WebPageProxy.cpp:16484
      - WebProcess JS executes globalThis.eval and serializes results per BiDi spec. File:
          - Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.js:49
          - Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.js:133
  - PrimitiveProtocolValue-compliant serialization
      - JS serializer maps primitives to {type, value} with special number handling (“NaN”, “Infinity”, “-Infinity”, “-0”), bigint as string. File:
          - Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.js:334
      - C++ deserializer converts JSON envelope to BidiScript.RemoteValue (accepts number or string for special number cases). File:
          - Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:58
  - Result and exception shaping
      - Success: returns realm and result (RemoteValue). Exceptions: builds ExceptionDetails including parsed stack frames from JS Error.stack. Files:
          - Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:333
  - Target precedence and realm mapping
      - context takes precedence over realm; simple realm<->context stub used to fill result.realm. Files:
          - Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:214
  - Input validation and error mapping
      - Strict validation for target, resultOwnership, and serializationOptions (type/enum range). Files:
          - Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:288
          - Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:72
      - Error code mapping aligned to BiDi: window → “no such browsing context”, frame → “no such frame”. File:
          - Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:158
  - Protocol schema updates
      - Added primitive types to RemoteValueType, fixed includeShadowTree name. File:
          - Source/WebKit/UIProcess/Automation/protocol/BidiScript.json:108
          - Source/WebKit/UIProcess/Automation/protocol/BidiScript.json:131
  - Test expectations
      - Unskipped the four targeted tests; narrowed evaluate.py skips. File:
          - WebDriverTests/TestExpectations.json:3235

Bug URL: https://bugs.webkit.org/show_bug.cgi?id=301159

Reviewed by NOBODY (OOPS!).

Why The Tests Pass

  - invalid.py
      - Invalid types/values now throw “invalid argument”; unknown context/realm maps to “no such frame”; invalid resultOwnership is rejected up-front. Files: BidiScriptAgent.cpp:214,
        288, WebDriverBidiProcessor.cpp:72, 142–161.
  - primitive_values.py
      - Serializer/deserializer match the spec: correct {type,value} for strings, booleans, null, undefined, bigint-as-string, and special numbers (“NaN”, “Infinity”, “-Infinity”,
        “-0”). Files: WebAutomationSessionProxy.js:334, BidiScriptAgent.cpp:58.
  - strict_mode.py
      - Expressions honor 'use strict' semantics during evaluation; first assignment to undeclared var throws with proper ExceptionDetails, subsequent non‑strict creates the global and
        strict re-assignment succeeds. Files: WebAutomationSessionProxy.js:133, 334, BidiScriptAgent.cpp:333.
  - target.py
      - context takes precedence over realm, and the response includes the effective realm with the serialized result. File: BidiScriptAgent.cpp:214.

No new tests (OOPS!).

* Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
(WebKit::convertJSONToRemoteValue):
(WebKit::BidiScriptAgent::evaluate):
(WebKit::BidiScriptAgent::finishEvaluateBidiScriptResult):
(WebKit::BidiScriptAgent::serializeAsRemoteValue):
(WebKit::BidiScriptAgent::RealmRegistryStub::realmIdForContext const):
(WebKit::BidiScriptAgent::RealmRegistryStub::contextForRealmId const):
* Source/WebKit/UIProcess/Automation/BidiScriptAgent.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::evaluateBidiScript):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::processBidiMessage):
(WebKit::toBidiErrorCode):
* Source/WebKit/UIProcess/Automation/protocol/BidiScript.json:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::evaluateBidiScript):
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.h:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.js:
(let.AutomationSessionProxy.prototype.evaluateBidiScript):
(let.AutomationSessionProxy.prototype._executeBidiScript):
(let.AutomationSessionProxy):
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.messages.in:
* WebDriverTests/TestExpectations.json:

lauromoura commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -35,12 +35,16 @@
 #include "WebDriverBidiProtocolObjects.h"
 #include "WebPageProxy.h"
 #include <wtf/TZoneMallocInlines.h>
+#include <cmath>
```
Shouldn't `cmath` come before the previous line? (Alphabetical ordering)

Strange that `check-webkit-style` did not warn about it.

lauromoura commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -110,36 +211,488 @@ void BidiScriptAgent::callFunction(const String& functionDeclaration, bool await
+
+                auto callFrames = JSON::ArrayOf<BidiScript::StackFrame>::create();
+
+                if (auto exceptionValue = envelopeObj->getValue("e"_s)) {
```
This code assembling the exception information should be refactored into separate methods (e.g. one to build the exception object, other to parse the strings, etc).

It'll make the code more readable, and ideally accessible to all agents, as it can be useful for other events/commands like `script.callFunction`, `log.entryAdded`.

lauromoura commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -51,6 +55,103 @@ BidiScriptAgent::BidiScriptAgent(WebAutomationSession& session, BackendDispatche
 
 BidiScriptAgent::~BidiScriptAgent() = default;
 
+static RefPtr<Inspector::Protocol::BidiScript::RemoteValue> convertJSONToRemoteValue(const JSON::Value* jsonValue)
```
The name of this function is kinda confusing against the other `BidiScriptAgent::serializeAsRemoteValue`.

IIUC, this function here is called just to convert back the `RemoteValue` JSON returned by the JS wrappers (which is already in the `type`+`value` format), while `serializeAsRemoteValue` implements the algorithm of the same name, converting actual "raw" JS values into `RemoteValue`, right?

lauromoura commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -51,6 +55,103 @@ BidiScriptAgent::BidiScriptAgent(WebAutomationSession& session, BackendDispatche
+        return remoteValue;
+    }
+
+    // Fallback
```
Add a `FIXME` comment about the remaining types not yet implemented?

lauromoura commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -110,36 +211,488 @@ void BidiScriptAgent::callFunction(const String& functionDeclaration, bool await
+
     ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!browsingContext, InvalidParameter);
-    ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!session->webPageProxyForHandle(*browsingContext), WindowNotFound);
+    ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!session->webPageProxyForHandle(*browsingContext), FrameNotFound);
```
We should be able to call functions on child contexts, no? https://w3c.github.io/webdriver-bidi/#get-a-realm-from-a-navigable does not bubble up to the parent context. In this case, we should be ready to receive context handles associated with `"frame-UUID"` handles on the browser side (for which `webPageProxyForHandle` fails).

I've submitted a fix for this specific issue in https://github.com/WebKit/WebKit/pull/53658, but we can work on it in parallel.

juliopiubello replied:
We held back on the implementation here because we scoped it to be inside of a "getRealms" scope ticket. We were afraid to increase the size of the PR too much, so we tried at maximum to split it on different PRs.

We would keep pushing progressive PRs that would increase the scripts section features

lauromoura commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -110,36 +211,488 @@ void BidiScriptAgent::callFunction(const String& functionDeclaration, bool await
+        // Proper validation should happen in the dispatcher/generator.
+    }
+
+    // Validate serialization options if provided
```
Comment not needed.

lauromoura commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -110,36 +211,488 @@ void BidiScriptAgent::callFunction(const String& functionDeclaration, bool await
+    if (optionalSerializationOptions) {
+        auto& serializationOptions = *optionalSerializationOptions;
+
+        // Validate maxDomDepth type and value
```
Ditto

lauromoura commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -110,36 +211,488 @@ void BidiScriptAgent::callFunction(const String& functionDeclaration, bool await
+            ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(std::floor(maxDomDepthValue) != maxDomDepthValue, InvalidParameter);
+        }
+
+        // Validate maxObjectDepth type and value
```
Ditto.

lauromoura commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -110,36 +211,488 @@ void BidiScriptAgent::callFunction(const String& functionDeclaration, bool await
+            ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(std::floor(maxObjectDepthValue) != maxObjectDepthValue, InvalidParameter);
+        }
+
+        // Validate includeShadowTree type and value
```
Ditto.

lauromoura commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -110,36 +211,488 @@ void BidiScriptAgent::callFunction(const String& functionDeclaration, bool await
+        if (auto maxObjectDepthJSON = optionalSerializationOptions->getValue("maxObjectDepth"_s)) {
+            double depthValue;
+            if (maxObjectDepthJSON->asDouble(depthValue) && depthValue >= 0 && std::floor(depthValue) == depthValue)
+                maxObjectDepth = static_cast<int>(depthValue);
```
Should we assign this var in the checks above (~295)? To avoid re-doing the look-up and checks again here.

lauromoura commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -110,36 +211,488 @@ void BidiScriptAgent::callFunction(const String& functionDeclaration, bool await
+        }
+    }
+
+    std::optional<double> callbackTimeout = std::nullopt; // Use default timeout
```
Comment not needed.

lauromoura commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -110,36 +211,488 @@ void BidiScriptAgent::callFunction(const String& functionDeclaration, bool await
 }
 
+RefPtr<Inspector::Protocol::BidiScript::RemoteValue> BidiScriptAgent::serializeAsRemoteValue(RefPtr<JSON::Value> value)
+{
```
It would be good to add a link to the spec: https://w3c.github.io/webdriver-bidi/#serialize-as-a-remote-value

lauromoura reviewed (changes requested):
Thanks for the PR! Testing locally, the results seem to be ok. Here's an initial review of some C++ bits.

Commit:
refactor(bidi): factor exception handling; rename RemoteValue helper

  - Extract parseStackTrace() and buildExceptionDetailsFromExceptionValue()
  - Rename convertJSONToRemoteValue() to deserializeRemoteValue()
  - Validate and use serializationOptions.maxObjectDepth; remove duplicate parsing
  - Document “serialize as a remote value” algorithm; add FIXME for unhandled types
  - Reorder BidiScript.json RemoteValue enum entries to match spec order

Commit:
fix(automation/bidi): make ParsedStackTrace default-constructible

  Initialize ParsedStackTrace::callFrames in a default constructor to fix a build error (non-default-constructible WTF::Ref). Remove redundant initializations now covered by the ctor

Commit:
refactor(automation): inline StackFrame addItem to satisfy WebKit guidelines

  - Replace temporary Ref with inline addItem(...release()) in parseStackTrace.
  - Eliminates “Uncounted call argument for a raw pointer/reference parameter” warning on mac CI.
  - No functional change; ownership transfer is explicit

Commit:
fix: mac-safer-cpp static analyzer errors in BidiScriptAgent

The UncountedCallArgsChecker flagged several patterns where builder results
were passed directly to methods without first being stored in a local variable.

Commit:
fix: Refactor exception handling and fix additional mac-safer-cpp issues in BidiScriptAgent

This commit refactors exception handling in BidiScriptAgent to reduce code
duplication and addresses additional mac-safer-cpp static analyzer warnings
by using more explicit type declarations.

Changes include:
- Added createEmptyStackTrace() helper to create empty stack traces
- Added createExceptionDetails() helper to standardize exception detail creation
- Replaced auto with explicit types (RefPtr<>, Ref<>, String) for clarity
- Reduced repetitive exception handling code in callFunction() and finishEvaluateBidiScriptResult()

Commit:
fix(bidi): drop unused bindings

Commit:
Refactor stack frame creation in BidiScriptAgent to streamline code and improve readability

Commit:
fix: UncountedCallArgsChecker: Store result.callFrames in a local variable before calling addItem() to avoid passing an uncounted reference as a call argument
